runtime/near-vm-logic/src/vmstate.rs

```diff
@@ -0,0 +1,177 @@
+    }
+
+    /// Copies data from guest memory into provided buffer accounting for gas.
+    fn get_into(&self, gas_counter: &mut GasCounter, offset: u64, buf: &mut [u8]) -> Result<()> {
+        gas_counter.pay_base(read_memory_base)?;
+        let len = u64::try_from(buf.len()).map_err(|_| HostError::MemoryAccessViolation)?;
+        gas_counter.pay_per(read_memory_byte, len)?;
+        self.0.read_memory(offset, buf).map_err(|_| HostError::MemoryAccessViolation.into())
+    }
+
+    /// Copies data from guest memory into a newly allocated vector accounting for gas.
+    pub(super) fn get_vec(
+        &self,
+        gas_counter: &mut GasCounter,
+        offset: u64,
+        len: u64,
+    ) -> Result<Vec<u8>> {
+        gas_counter.pay_base(read_memory_base)?;
+        gas_counter.pay_per(read_memory_byte, len)?;
+        self.get_vec_for_free(offset, len)
+    }
+
+    /// Like [`Self::get_vec`] but does not pay gas fees.
+    pub(super) fn get_vec_for_free(&self, offset: u64, len: u64) -> Result<Vec<u8>> {
+        // This check is redundant in the sense that read_memory will perform it
+        // as well however it’s here to validate that `len` is a valid value.
+        // See documentation of MemoryLike::read_memory for more information.
+        self.0.fits_memory(offset, len).map_err(|_| HostError::MemoryAccessViolation)?;
+        let mut buf = vec![0; len as usize];
+        self.0.read_memory(offset, &mut buf).map_err(|_| HostError::MemoryAccessViolation)?;
+        Ok(buf)
+    }
+
+    /// Copies data from provided buffer into guest memory accounting for gas.
+    pub(super) fn set(
+        &mut self,
+        gas_counter: &mut GasCounter,
+        offset: u64,
+        buf: &[u8],
+    ) -> Result<()> {
+        gas_counter.pay_base(write_memory_base)?;
+        gas_counter.pay_per(write_memory_byte, buf.len() as _)?;
+        self.0.write_memory(offset, buf).map_err(|_| HostError::MemoryAccessViolation.into())
+    }
+
+    memory_get!(u128, get_u128);
+    memory_get!(u32, get_u32);
+    memory_get!(u16, get_u16);
+    memory_get!(u8, get_u8);
+    memory_set!(u128, set_u128);
+}
+
+/// Registers to use by the guest.
+///
+/// Provides interface to access registers while correctly accounting for gas
+/// usage.
+///
+/// See documentation of [`Memory`] for more motivation for this struct.
+#[derive(Default, Clone)]
+pub(super) struct Registers(std::collections::HashMap<u64, Box<[u8]>>);
+
+impl Registers {
+    /// Returns register with given index.
+    ///
+    /// Returns an error if (i) there’s not enough gas to perform the register
+    /// read or (ii) register with given index doesn’t exist.
+    pub(super) fn get<'s>(
+        &'s self,
+        gas_counter: &mut GasCounter,
+        register_id: u64,
+    ) -> Result<&'s [u8]> {
+        if let Some(data) = self.0.get(&register_id) {
+            gas_counter.pay_base(read_register_base)?;
+            gas_counter.pay_per(read_register_byte, data.len() as u64)?;
```

Comment:
Would it make sense to apply the same treatment to the conversion as done in the `Memory::get_into`?

```suggestion
            let len = u64::try_from(data.len()).map_err(|_| HostError::MemoryAccessViolation)?;
            gas_counter.pay_per(read_register_byte, len)?;
```

In practice this should be always checked by the `set` but consistency can’t hurt :)